File: app/services/api/src/dtos/input/activity.py
from typing import Optional, List

from pydantic import BaseModel, Field

from app.libs.db import models


class CreateActivity(BaseModel):
    """
    DTO for creating a new activity
    """
    institution_id: str
    course_id: str
    activity_type: models.ActivityType
    group_id: str
    professor_id: Optional[str] = None
    duration_slots: int = Field(default=2, gt=0)
    required_room_features: List[str] = Field(default_factory=list)
    frequency: models.Frequency = models.Frequency.WEEKLY
    selected_timeslot: Optional[models.SelectedTimeslot] = None


class UpdateActivity(BaseModel):
    """
    DTO for updating an existing activity
    """
    course_id: Optional[str] = None
    activity_type: Optional[models.ActivityType] = None
    group_id: Optional[str] = None
    professor_id: Optional[str] = None
    duration_slots: Optional[int] = None
    required_room_features: Optional[List[str]] = None
    frequency: Optional[models.Frequency] = None
    selected_timeslot: Optional[models.SelectedTimeslot] = None

File: app/services/ui/src/types/activity.ts
export interface ActivitySelectedTimeslot {
  start_timeslot: number;
  active_weeks: number[];
}

export interface ActivityData {
  id?: string;
  _id?: string;
  institution_id: string;
  course_id: string;
  activity_type: string;
  group_id: string;
  professor_id?: string | null;
  duration_slots: number;
  required_room_features?: string[];
  frequency: string;
  selected_timeslot?: ActivitySelectedTimeslot | null;
}

export interface ActivityShape {
  id: string;
  institution_id: string;
  course_id: string;
  activity_type: string;
  group_id: string;
  professor_id?: string | null;
  duration_slots: number;
  required_room_features: string[];
  frequency: string;
  selected_timeslot?: ActivitySelectedTimeslot | null;
}

export class Activity implements ActivityShape {
  id: string;
  institution_id: string;
  course_id: string;
  activity_type: string;
  group_id: string;
  professor_id?: string | null;
  duration_slots: number;
  required_room_features: string[];
  frequency: string;
  selected_timeslot?: ActivitySelectedTimeslot | null;

  constructor(data: ActivityData) {
    this.id = data.id ?? data._id ?? '';
    this.institution_id = data.institution_id;
    this.course_id = data.course_id;
    this.activity_type = data.activity_type;
    this.group_id = data.group_id;
    this.professor_id = data.professor_id ?? null;
    this.duration_slots = data.duration_slots;
    this.required_room_features = data.required_room_features ?? [];
    this.frequency = data.frequency;
    this.selected_timeslot = data.selected_timeslot ?? null;
  }

  static from(data: ActivityData): Activity {
    return new Activity(data);
  }
}

